### DISCUSSION TITLE
The volume scaling is inconsistent with Windows's (linear vs. cubic, rather important!)

### DISCUSSION THREAD

**Discussion by Percitomi** [File-New-Project/EarTrumpet#861] [Completed ideas]
I am glad this app exists.

However, the volume sliders in it scale voltage linearly. This results in a barely audible 3 decibel difference between settings 50 and 100 on the slider. Not only is this far from ideal (human hearing uses an exponential scale), but is also inconsistent with the noticeably better scaling (not sure what the exact formula is, it seems to be close to cubic) Windows itself uses for its master volume slider. This inconsistency means that using 10 master volume with 10 EarTrumpet volume is very different from using 100 master volume with 1 EarTrumpet volume (about 20 decibels of difference in this case, a huge difference!).

If it's possible to implement (maybe it's just changing a single formula in a single function somewhere, I don't know), I would really strongly appreciate if the app also used cubic scaling. It would provide better consistency and better volume control.

Alternatively, if there's a place where I myself could put a custom formula into the source to have a non-linear scale, let me know.

**Comment by riverar** [comment]
@Percitomi Added logarithmic scale, give it a whirl in dev builds.

**Reply by BuongiornoTexas** [reply]
Hi @riverar - This looks to be a great idea. I have run into a minor disconnect using hotkeys - I can use Ear Trumpet to set a per device or per app volume with hotkeys in the current volume scale. But if I want to change volume, I have to use something like powertoys, which adjusts windows base volume in 2% increments. So if I start from 0 volume in the log volume, the ear trumpet volume runs in the following increments: 0-32-44-51-56 ... (corresponding to windows volumes 0-2-4-6- ...).

This makes sense for what the volume is actually doing, but breaks my perception of a smooth volume change.

A possible enhancement to address this could be allowing Ear Trumpet actions to perform fixed increment volume changes on the ear trumpet scale (i.e. effectively duplicate the existing windows increment capability, but applied to the active volume scale). I don't think this is any sort of priority for dev effort, but without it, log volume scales may be a power user function only.

**Reply by riverar** [reply]
Hey @BuongiornoTexas, not sure I'm following. I'll write out how I think it should work and maybe you can help me understand what we're missing.

With this new feature enabled, all volume adjustments in EarTrumpet are log scale. External adjustments (e.g. sndvol.exe) happen in linear scale. Both are tied together and reflect the correct values in their respective scales. EarTrumpet hotkeys should increment/decrement in the configured scale. External hotkeys will increment/decrement on the linear scale. EarTrumpet's +2 increment is currently still +2 in all scales, although it has been suggested that might be too much for log scale. (Should we decrease that to 1?)

**Reply by BuongiornoTexas** [reply]
> EarTrumpet hotkeys should increment/decrement in the configured scale. 

Ah! I think I've fallen foul of this before.

TLDR: Ear Trumpet is doing what is should, but I can't use its hotkey volume control because it is not per device. 

(I had forgotten Ear Trumpet does implement volume increment/decrement.)

If I have this right, this almost sorts my issue - unfortunately, I think I get caught out by Ear Trumpet forcing increment/decrement on all devices. For the way my system is set up, I have several devices that I want to keep at 100% at all times (mainly amps/speakers), for which I've got a calibrated maximum volume setup - for these devices I typically tweak final volume in app. 

However, my headphones are for general listening and because of the vagaries of the many sources I put through them, I want to be able to increment/decrement the volume on these independently of all other devices. Which I can do with Powertoys, but if I understand correctly I don't think I can do with Ear Trumpet? 

And if I use Powertoys per device with Ear Trumpet logscale, I get correct, but wonky looking behaviour. 

Ideally, I'd prefer to do this in Ear Trumpet with per device volume shortcuts - but as noted above, I think I'm in a small minority for this use case.

**Reply by riverar** [reply]
@BuongiornoTexas If you install the dev builds, you'll get access to EarTrumpet Actions which lets you build up something more advanced, pictured below. 🙃

![image](https://user-images.githubusercontent.com/475132/232277385-019fe42b-6a7e-4299-94cf-beed40ad6288.png)

**Reply by BuongiornoTexas** [reply]
And now we can colour me an idiot. An embarrassed idiot! 

I am already using the dev build and an action to allow me to reset my headphones to a volume level. My mistake was assuming that set was the only action allowed for volume (apparently I can't interpret blue underlines consistently in my old age).

So, yes, I very much think this will sort out my problem! I've set it up and am giving it try now. Thanks again for the great tool and the time you spend responding on it!

As the action allows me to set the increment manually, I'll give 1 and 2 a try and get back with experience in a few weeks.

The one thing that would still be a nice to have is a flyout showing the current volume after the volume change (the windows flyout shows briefly after using a Powertoys increment/decrement). I mentioned this in discussion #1154. 

An alternative implementation could be a new Ear Trumpet Action to show an Ear-Trumpet flyout for a user configurable period? Users could then add it to their action chain if they wanted this information?

**Reply by BuongiornoTexas** [reply]
@riverar - As promised, I've spent a couple of weeks using the log scale. I've found the default increment/decrement of 2 to be fine. Going down to 1 would probably be too fine grained. 

And thanks again for highlight the excellent functions of Ear Trumpet - it is now my only tool. 

I will re-iterate my request above - could we add an ear trumpet action to pop up a flyout/pop up volume indicator to provide a visual confirmation of volume level? Happy to put this in as a feature request. Unfortunately, implementation is probably outside my skill set - I struggle enough with python alone.

**Reply by riverar** [reply]
Thanks for reporting back, greatly appreciated. There should be a feature request for this already to upvote, it sounds familiar. If you don't find one, please do create one and tag it here.

So if we did add this, where would you like it to appear? Perhaps centered and offset from the taskbar like Windows does when media buttons are pressed?

**Reply by Percitomi** [reply]
Thank you a lot for the feature. It's great.
What I've noticed is that step size for master volume and app volume is different.
It seems to be 0.75 dB for master volume and 0.5 dB for app volume, which is consistent with Windows using x³ and x² respectively. It kind of bugs me that even Windows has a mismatch between the two....
Are 0.75dB and 0.5dB correct?

**Reply by riverar** [reply]
@Percitomi Hmm. We use [IAudioEndpointVolume::SetMasterVolumeLevelScalar](https://learn.microsoft.com/en-us/windows/win32/api/endpointvolume/nf-endpointvolume-iaudioendpointvolume-setmastervolumelevelscalar) and [ISimpleAudioVolume::SetMasterVolume](https://learn.microsoft.com/en-us/windows/win32/api/audioclient/nf-audioclient-isimpleaudiovolume-setmastervolume) internally. I think you're saying the log values we're passing in are being interpreted differently on the Windows side--correct? Maybe that's documented somewhere, I'll take a look.

**Comment by BuongiornoTexas** [comment]
> There should be a feature request for this already to upvote, it sounds familiar

That was probably #1154, which I raised a while back and I thought was a bit more general - after re-reading, you are correct - I've already asked for this (or at least something very close to it).

> So if we did add this, where would you like it to appear?

I'd suggest either the top and center of screen, as that's where my visual focus tends to go, or as a brief pop out from the system tray on the basis of convention. Though, I'd be happy with anywhere as long as I got the visual feedback.

**Comment by Gravifer** [comment]
Maybe we may add an option to show -dB values instead of 1-100 numerics for the logarithmic scale, more like a fader in DAWs? I'll try to PR this if I got a clue.

**Comment by Gravifer** [comment]
My request and requests for capped-renormalizing like #1273 can both be more generally termed as request for more general scaling-functions (paired with a prescriped inverse map of it). I'll try to navigate through the codebase for this.
